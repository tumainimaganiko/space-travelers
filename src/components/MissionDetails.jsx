import { useDispatch } from 'react-redux';
import { useState } from 'react';
import PropTypes from 'prop-types';
import styles from '../styles/Mission.module.css';
import { joinMission, leaveMission } from '../redux/missions/missionsSlice';

const MissionDetails = ({
  name, description, id, reserved,
}) => {
  const myStatus = {
    backgroundColor: reserved ? '#379cf6' : '#36454F',
    display: 'block',
    width: '120px',
    border: '1px solid gray',
    color: 'white',
    alignSelf: 'center',
  };
  const styleRow = {
    backgroundColor: reserved ? 'white' : '#E5E4E2',
  };

  const dispatch = useDispatch();
  const [text, setText] = useState('Not a Member');

  const handleJoin = (e) => {
    e.preventDefault();
    dispatch(joinMission(id));

    if (!reserved) {
      setText('Active Member');
    }
  };
  const handleLeave = (e) => {
    e.preventDefault();
    dispatch(leaveMission(id));
    if (reserved) {
      setText('Not a member');
    }
  };

  return (
    <tr style={styleRow}>
      <td className={styles.name}>{name}</td>
      <td className={styles.description}>{description}</td>
      <td>
        <span style={myStatus}>
          {' '}
          {text}
        </span>
      </td>
      <td>
        {!reserved && (
          <button type="button" className={styles.joinBtn} onClick={handleJoin}>
            Join Mission
          </button>
        )}
        {reserved && (
          <button
            type="button"
            className={styles.leaveBtn}
            onClick={handleLeave}
          >
            Leave Mission
          </button>
        )}
      </td>
    </tr>
  );
};

MissionDetails.propTypes = PropTypes.shape({
  name: PropTypes.string,
  description: PropTypes.string,
  id: PropTypes.string,
  reserved: PropTypes.bool,
}).isRequired;

MissionDetails.defaultProps = {
  reserved: false,
};
export default MissionDetails;
