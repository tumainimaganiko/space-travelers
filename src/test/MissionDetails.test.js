import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import MissionDetails from '../components/MissionDetails';
import myMissions from './missionsData';

const reducer = (
  state = {
    missions: { missions: myMissions },
  },
) => state;

const store = configureStore({ reducer });

it('Should render Mission details', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <MissionDetails
          name="space 1"
          description="first space"
          id="firstItem"
          reserved
        />
      </Provider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
