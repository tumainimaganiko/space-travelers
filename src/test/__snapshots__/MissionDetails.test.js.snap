// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Mission details 1`] = `
<tr
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  <td
    className="name"
  >
    space 1
  </td>
  <td
    className="description"
  >
    first space
  </td>
  <td>
    <span
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#379cf6",
          "border": "1px solid gray",
          "color": "white",
          "display": "block",
          "width": "120px",
        }
      }
    >
       
      Not a Member
    </span>
  </td>
  <td>
    <button
      className="leaveBtn"
      onClick={[Function]}
      type="button"
    >
      Leave Mission
    </button>
  </td>
</tr>
`;
